### DISCUSSION TITLE
Question: How can store/middleware respond to lifecycle events?

### DISCUSSION THREAD

**Discussion by matthewjamesadam** [pmndrs/zustand#792] [Q&A]
Hello, thanks for making Zustand!  I have a quick question: what is the recommended way for Zustand stores (or middleware) to respond to store lifecycle events?  Here is our situation:

* We have a store whose purpose is to monitor some external agent, and publish state from that agent in the store's state.
* The store is only alive while certain UIs are active
* Monitoring the agent is somewhat expensive, so once the UI shuts down, the store should deactivate, and the monitoring should be shut down.

It looks like under the hood, `create` / `useStore` is calling `api.subscribe` / `api.unsubscribe`, which makes sense, but as far as I can tell there is no way for the store (or middleware) to react to the unsubscribe/destroy events, such that when the last subscriber is unsubscribed, the store can take some action.  Any suggestions on how I should deal with this scenario?  I can picture wrapping `create` in a function that adds some extra logic that will call lifecycle methods on the store, but I'm not sure if that's the right approach.

**Comment by dai-shi** [comment]
>  I can picture wrapping `create` in a function that adds some extra logic that will call lifecycle methods on the store, but I'm not sure if that's the right approach.

It sounds like a right approach. Do you mean wrapping `create` or a store? Middleware can do too, but typing middleware would be too hard.

Can you create a small code snippet to illustrate the idea? It might give some hints to get a better approach.
